plugins/acl/dynamic-role-based/src/org/apache/cloudstack/acl/DynamicRoleBasedAPIAccessChecker.java

```diff
@@ -0,0 +1,170 @@
+        if (permissions == null || permissions.isEmpty() || Strings.isNullOrEmpty(commandName)) {
+            return false;
+        }
+        for (final RolePermission permission : permissions) {
+            if (permission.getPermission() != permissionToCheck) {
+                continue;
+            }
+            try {
+                final Rule rule = new Rule(permission.getRule());
+                if (rule.matches(commandName)) {
+                    return true;
+                }
+            } catch (InvalidParameterValueException e) {
+                LOGGER.warn("Invalid rule permission, please fix id=" + permission.getId() + " rule=" + permission.getRule());
+                continue;
+            }
+        }
+        return false;
+    }
+
+    public boolean isDisabled() {
+        return !roleService.isEnabled();
+    }
+
+    @Override
+    public boolean checkAccess(User user, String commandName) throws PermissionDeniedException {
+        if (isDisabled()) {
+            return true;
+        }
+        Account account = accountService.getAccount(user.getAccountId());
+        if (account == null) {
+            throw new PermissionDeniedException("The account id=" + user.getAccountId() + "for user id=" + user.getId() + "is null");
+        }
+
+        final Role accountRole = roleService.findRole(account.getRoleId());
+        if (accountRole == null || accountRole.getId() < 1L) {
+            denyApiAccess(commandName);
+        }
+
+        // Allow all APIs for root admins
+        if (accountRole.getRoleType() == RoleType.Admin && accountRole.getId() == RoleType.Admin.getId()) {
+            return true;
+        }
+
+        final List<RolePermission> rolePermissions = roleService.findAllPermissionsBy(accountRole.getId());
+
+        // Check for allow rules
+        if (checkPermission(rolePermissions, RolePermission.Permission.ALLOW, commandName)) {
+            return true;
+        }
+
+        // Check for deny rules
+        if (checkPermission(rolePermissions, RolePermission.Permission.DENY, commandName)) {
+            denyApiAccess(commandName);
+        }
+
+        // Check annotations
+        if (annotationRoleBasedApisMap.get(accountRole.getRoleType()) != null
+                && annotationRoleBasedApisMap.get(accountRole.getRoleType()).contains(commandName)) {
+            return true;
+        }
+
+        denyApiAccess(commandName);
+        return false;
+    }
+
+    public void addApiToRoleBasedAnnotationsMap(final RoleType roleType, final String commandName) {
+        if (roleType == null || Strings.isNullOrEmpty(commandName)) {
+            return;
+        }
+        final Set<String> commands = annotationRoleBasedApisMap.get(roleType);
+        if (commands != null && !commands.contains(commandName)) {
+            commands.add(commandName);
```

Comment:
@koushik-das (3) from step 2, this methods populates an internal map that is a list of authorized apis for a given role types; we've only 4 role types used across cloudstack (and also in commands.properties aka the static checker) namely admin, resource admin, domain admin and user